lib/db/observed.go

```diff
@@ -111,6 +111,41 @@ func (db *Lowlevel) RemovePendingFolder(id string) {
 	}
 }
 
+func (db *Lowlevel) RemovePendingFoldersBeforeTime(device protocol.DeviceID, oldest time.Time) ([]string, error) {
+	prefixKey, err := db.keyer.GeneratePendingFolderKey(nil, device[:], nil)
+	if err != nil {
+		return nil, err
+	}
+	iter, err := db.NewPrefixIterator(prefixKey)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Release()
+	oldest = oldest.Round(time.Second)
+	var res []string
+	for iter.Next() {
+		var bs []byte
+		var of ObservedFolder
+		if bs, err = db.Get(iter.Key()); err != nil {
+			l.Infof("Invalid pending folder entry, deleting from database: %x", iter.Key())
+		} else if err = of.Unmarshal(bs); err != nil {
+			l.Infof("Invalid pending folder entry, deleting from database: %x", iter.Key())
+		} else if of.Time.Before(oldest) {
+			l.Infof("Removing stale pending folder %v from device %s, last seen %v",
+				of.Label, device.Short(), of.Time)
+		} else {
+			// Keep entries younger or equal to the given timestamp
+			continue
+		}
+		if err := db.Delete(iter.Key()); err != nil {
+			l.Warnf("Failed to remove pending folder entry: %v", err)
+		}
+		folderID := string(db.keyer.FolderFromPendingFolderKey(iter.Key()))
+		res = append(res, folderID)
```

Comment:
I think we only want to append in the "expired" case?
Or do we really want to report folders that we can't read/unmarshal/etc as removed?
I guess in theory they are removed, but if Get(iter.Key()) failed, not sure what we are removing?

Reply:
>  Or do we really want to report folders that we can't read/unmarshal/etc as removed?

Hard question.  It should never make a difference because every time the current list is requested through `PendingFolders()`, that will already weed out invalid entries.  OTOH, that happens without `cleanPending()` ever noticing so will also *not* be reported in the event.

For the sake of lower code complexity I'd prefer the current approach.  If it ever reports something as removed which the GUI didn't have in its cached list, that will fail gracefully.  It's also our only chance to report this over the event API and certainly doesn't hurt for keeping the consumer's cache in sync.

Reply:
At the very least it's wrong the report something as removed when `Delete` failed. I'd also tend to only report truly expired entries. And doesn't feel any more complex to move the two lines above into the second if block above.

Reply:
You're right about failed deletions.  That problem isn't solved by moving the `append()` call before the actual `Delete()` though.  I conditionalized it now based on the folderID variable, which is not populated for invalid entries.